Source/DafnyCore/Verifier/Translator.DataTypes.cs

```diff
@@ -16,7 +16,11 @@ partial class Translator {
       Contract.Requires(dt != null);
       Contract.Requires(sink != null && predef != null);
 
-      var constructorFunctions = dt.Ctors.ToDictionary(ctor => ctor, ctor => AddDataTypeConstructor(dt, ctor));
+      var mayInvolveReferences = UserDefinedType.FromTopLevelDecl(dt.tok, dt).MayInvolveReferences;
+      var constructorFunctions = dt.Ctors.ToDictionary(ctor => ctor, ctor => AddDataTypeConstructor(dt, ctor, mayInvolveReferences));
+      if (!mayInvolveReferences) {
+        AddCommonIsAllocConstructorAxiom(dt);
```

Comment:
Why does this work? Consider the value Cons(3,Cons(4,Nil)). Does it work because if Cons(3,Cons(4,Nil)) is allocated, then we can easily deduct that Cons(4,Nil) is as well? Or is it because it is always good enough to know that Cons(3,Cons(4,Nil)) is allocated?